Normalize labels before addition or removal

kailan opened the pull request:
The crates.io repository has some cute emoji in the label names, but this interferes with usage of label-related triagebot commands. Example:


<img width="937" height="596" alt="Screenshot 2025-07-29 at 12 19 18" src="https://github.com/user-attachments/assets/94d2ac45-ed1e-45e4-b452-888154f05aa8" />

It seems unreasonable to expect users to exactly match the label names in these cases, and it would be a shame to remove the emoji from the label names, so I feel that it makes sense to make triagebot a bit more lax when it comes to finding labels.

This PR adds a `normalize_and_match_labels` function, which takes a set of "requested" label names and attempts to match them to the labels available in the repository. This will always return an exact match, if it exists, but otherwise will strip emoji, ignore casing, and ignore leading and trailing whitespace in an attempt to find the relevant label.

This function is used in the lower level `add_labels` and `remove_label` functions so that it can apply across any triagebot commands that handle labels. This means that the existing configuration for autolabelling in crates.io will start working:

```toml
[autolabel."A-backend"]
trigger_files = [
    "src/",
]
```

This PR is somewhat related to #1372.

Urgau commented:
> It seems unreasonable to expect users to exactly match the label names in these cases

I feel like it's unreasonable for GitHub to allow emoji in label names but here we are.

> This PR adds a `normalize_and_match_labels` function

Unless I'm mistaken, your function would allow the same prefix to be match to multiple emoji. "label :shrug:", "label :+1:" and "label :100:" would all be matched by "label", right?

I don't think we should any of those to match "label", I feel like we should require perfect match in those cases.

-------

I didn't look too closely at the code but I saw that there wasn't any caching done, we'll need to add some, at minimum per webhook request, maybe in [`Issue`](https://github.com/rust-lang/triagebot/blob/0f75748b83675a6d9afa25f4fdac94944cbc6954/src/github.rs#L364) to start and in the long run per repository.

kailan commented:
> I feel like it's unreasonable for GitHub to allow emoji in label names but here we are.

Can't argue with that, lol.

> Unless I'm mistaken, your function would allow the same prefix to be match to multiple emoji. So "label 🤷", "label 👍" and "label 💯" would all be matched by "label", right?
> 
> I don't think we should allow any of those to match "label", I feel like we should require perfect match in those cases.

This is a great point, so I have added an error for this case:

`AmbiguousLabelMatch` –  "Unsure which label to use for `label` - could be one of: `label 🤷`, `label 👍`, `label 💯`"

> I didn't look too closely at the code but I saw that there wasn't any caching done

I didn't implement any caching, however this implementation will send far fewer requests to GitHub – previously a `GET /labels/label_name` would have been made for every requested label, this is now replaced by a single `GET /labels` which provides the data needed to check all labels.

kailan commented on src/github.rs:
```diff
@@ -730,8 +745,71 @@ impl Issue {
+                .into());
+            } else {
+                found_labels.push(&found.first().unwrap().name);
+            }
```
TIL! this is great

Urgau commented on src/labels.rs:
Triagebot handles multiple services (mainly GitHub and Zulip) and this file while technically having being independent of GitHub has some assumptions about it, and doesn't make sense for Zulip.

Could you move it under `src/github/`?

Urgau replied:
Otherwise this looks good.

Urgau commented:
Thanks for the PR. Let's try it.

Urgau commented:
Let's do some quick sanity checks.

@rustbot label A-test

Urgau commented:
@rustbot label A-test-emoji

Urgau commented:
@rustbot label "A-test-emoji :smile:"

Urgau commented:
@rustbot label "-A-test-emoji :smile:" "-A-test"

Urgau commented:
@rustbot label A-test-emoji-2

kailan commented:
> Hum, this doesn't work 100% as I would have expected. GitHub allows both Unicode emoji characters and the Markdown named ones (`:smile:`, ...), and returns the raw/unexpected version in it's API.

Oh, that's definitely an oversight that should be addressed.

> I think this may have broken standard ascii labels in some way too; libc's autolabel has started claiming everything is unknown

even worse! Okay, I will revisit this with some new test cases based on these issues.

Probably worth rolling this back if you haven't already. Apologies for the inconvenience.

Urgau commented:
Revert on going at https://github.com/rust-lang/triagebot/pull/2133, apologies for the inconvenience.
